src/main/java/com/google/devtools/build/lib/rules/repository/RepoRecordedInput.java

```diff
@@ -106,121 +137,192 @@ public int compareTo(RepoRecordedInput o) {
+    }
 
-    public LabelFile(Label label) {
-      this.label = label;
+    @Override
+    public int hashCode() {
+      return Objects.hash(repoName, pathFragment);
     }
 
     @Override
-    String toStringInternal() {
-      return label.toString();
+    public String toStringInternal() {
+      // We store `@@foo//abc/def:ghi.bzl` as just `@@foo//abc/def/ghi.bzl`.
+      return repoName + "//" + pathFragment;
     }
 
     @Override
-    public SkyKey getSkyKey() {
-      return PackageLookupValue.key(label.getPackageIdentifier());
+    public SkyKey getSkyKey(Environment env) throws InterruptedException {
+      return repoName.isMain()
+          ? PrecomputedValue.PATH_PACKAGE_LOCATOR.getKey()
+          : RepositoryDirectoryValue.key(repoName);
     }
 
     @Override
     public boolean isUpToDate(Environment env, @Nullable String oldValue)
         throws InterruptedException {
-      PackageLookupValue pkgLookupValue = (PackageLookupValue) env.getValue(getSkyKey());
-      if (pkgLookupValue == null || !pkgLookupValue.packageExists()) {
-        return false;
+      Root root;
+      if (repoName.isMain()) {
+        root =
+            Root.fromPath(
+                PrecomputedValue.PATH_PACKAGE_LOCATOR.get(env).getPathEntries().get(0).asPath());
+      } else {
+        RepositoryDirectoryValue repoDirValue =
+            (RepositoryDirectoryValue) env.getValue(getSkyKey(env));
+        if (repoDirValue == null || !repoDirValue.repositoryExists()) {
+          return false;
+        }
+        root = Root.fromPath(repoDirValue.getPath());
       }
-      RootedPath rootedPath =
-          RootedPath.toRootedPath(pkgLookupValue.getRoot(), label.toPathFragment());
+      RootedPath rootedPath = RootedPath.toRootedPath(root, pathFragment);
       SkyKey fileKey = FileValue.key(rootedPath);
       try {
         FileValue fileValue = (FileValue) env.getValueOrThrow(fileKey, IOException.class);
         if (fileValue == null || !fileValue.isFile() || fileValue.isSpecialFile()) {
           return false;
         }
-        return oldValue.equals(RepositoryFunction.fileValueToMarkerValue(fileValue));
+        return oldValue.equals(fileValueToMarkerValue(fileValue));
       } catch (IOException e) {
         return false;
       }
     }
   }
 
   /**
-   * Represents a file input accessed during the repo fetch that is <i>not</i> addressable by a
-   * label. This most likely means that it's outside any known Bazel workspace.
+   * Represents a file input accessed during the repo fetch that is outside the current Bazel
+   * workspace. This file is addressed by its absolute path.
    */
-  public static final class AbsolutePathFile extends File {
-    final RootedPath path;
+  public static final class FileOutsideWorkspace extends File {
+    private final PathFragment path;
 
-    public AbsolutePathFile(RootedPath path) {
+    public FileOutsideWorkspace(PathFragment path) {
       this.path = path;
```

Comment:
nit: assert that this is absolute for peace of mind.

Reply:
done.